### PR TITLE
Improvements to dependents UI

### PR DESCRIPTION
**Pull request by davidackroyd99**
Addresses https://github.com/loic-sharma/BaGet/issues/570

Thanks for assigning this issue to me :) Decided to only return the information from the API that is actually required to build up the table, just to make it more performant. If you want me to change any of this I can 👍

### CODE REVIEW

File: src/BaGet.Core/Search/DependentsResponse.cs [hunk 1]
```diff
@@ -19,6 +19,36 @@ public class DependentsResponse
...
+    /// <summary>
+    /// A package that depends on the queried package.
+    /// </summary>
+    public class DependentResponse
+    {
+        /// <summary>
+        /// The dependent package key.
+        /// </summary>
+        [JsonPropertyName("key")]
+        public int Key { get; set; }
```
Comment (loic-sharma) [inline src/BaGet.Core/Search/DependentsResponse.cs:34]: What's the key for? Can we remove it?

File: src/BaGet.Core/Search/DependentsResponse.cs [hunk 2]
```diff
@@ -19,6 +19,36 @@ public class DependentsResponse
...
         /// </summary>
         [JsonPropertyName("data")]
-        public IReadOnlyList<string> Data { get; set; }
+        public IReadOnlyList<DependentResponse> Data { get; set; }
+    }
+
+    /// <summary>
+    /// A package that depends on the queried package.
+    /// </summary>
+    public class DependentResponse
```
Comment (loic-sharma) [inline src/BaGet.Core/Search/DependentsResponse.cs:28]: Let's call this class [`DependentResult`](https://github.com/loic-sharma/BaGet/blob/7c514bfe57a59b7ed086e37b73d61a8194306a20/src/BaGet.Protocol/Models/SearchResponse.cs#L11) to mirror the `SearchResponse`'s [`SearchResult`](https://github.com/loic-sharma/BaGet/blob/7c514bfe57a59b7ed086e37b73d61a8194306a20/src/BaGet.Protocol/Models/SearchResult.cs#L12).

File: src/BaGet.Core/Search/DatabaseSearchService.cs [hunk 3]
```diff
@@ -125,7 +125,12 @@ public async Task<DependentsResponse> FindDependentsAsync(string packageId, Canc
                 .OrderByDescending(p => p.Downloads)
                 .Where(p => p.Dependencies.Any(d => d.Id == packageId))
                 .Take(20)
-                .Select(p => p.Id)
+                .Select(r => new DependentResponse() {
```
Comment (loic-sharma) [inline src/BaGet.Core/Search/DatabaseSearchService.cs:128]: Please put `{` on a new line to follow the .NET Foundation's style:

```suggestion
                .Select(r => new DependentResponse
                {
```

(It's not very compact, I know 😅)

File: src/BaGet.UI/src/DisplayPackage/Dependents.tsx [hunk 4]
```diff
@@ -40,6 +50,25 @@ class Dependents extends React.Component<IDependentsProps, IDependentsState> {
     }).catch((e) => console.log("Failed to load dependents.", e));
   }
 
+  private getDependentsMessage() {
+    const hits = this.state.totalHits ?? -1;
+    const packageId = this.props.packageId;
+
+    if(hits < 0) {
```
Comment (loic-sharma) [inline src/BaGet.UI/src/DisplayPackage/Dependents.tsx:57]: Please add a space between `if` and `(`. Same comment for lines 61 and 65

File: src/BaGet.UI/src/DisplayPackage/Dependents.tsx [hunk 5]
```diff
@@ -49,19 +78,37 @@ class Dependents extends React.Component<IDependentsProps, IDependentsState> {
...
+                <tr>
+                  <th className="col-sm-10">Package</th>
+                  <th className="col-sm-2">Downloads</th>
+                </tr>
+              </thead>
+              <tbody>
+                {this.state.data.map(dependent => (
+                  <tr key={dependent.id}>
+                    <td>
+                      <a href={config.apiUrl + "/packages/" + dependent.id}>{dependent.id}</a>
```
Comment (loic-sharma) [inline src/BaGet.UI/src/DisplayPackage/Dependents.tsx:100]: Think can be:

```suggestion
                      <Link to {`/packages/${dependent.id}`}>{dependent.id}</Link>
```

For example, see how search results link to the package page: https://github.com/loic-sharma/BaGet/blob/7c514bfe57a59b7ed086e37b73d61a8194306a20/src/BaGet.UI/src/SearchResults.tsx#L198

File: src/BaGet.UI/src/DisplayPackage/Dependents.tsx [hunk 6]
```diff
@@ -40,6 +50,25 @@ class Dependents extends React.Component<IDependentsProps, IDependentsState> {
...
+    if(hits < 0) {
+      return ""
+    }
+
+    if(hits === 0) {
+      return `No packages depend on ${packageId}.`;
+    }
+
+    if(hits < 20) {
+      return `Showing ${hits} packages that depend on ${packageId}.`;
```
Comment (loic-sharma) [inline src/BaGet.UI/src/DisplayPackage/Dependents.tsx:66]: This message shouldn't be plural if there's only a single result. What do you think of just always showing the `Showing the top 20 packages that depend` if there are 1 or more hits, even if there are more than 20?

**Comment by loic-sharma** [comment]
This is awesome! I left a few minor comments, I'll merge this in once you address them.

**Comment by loic-sharma** [comment]
By the way, could you also rename this heading to `Used By` to match nuget.org?

https://github.com/loic-sharma/BaGet/blob/9c16b0068576495f647cc77c1c000124134444b2/src/BaGet.UI/src/DisplayPackage/DisplayPackage.tsx#L254

File: src/BaGet.Core/Search/DependentsResponse.cs (hunk 1 above)
Comment (davidackroyd99) [reply src/BaGet.Core/Search/DependentsResponse.cs:34]: I guess we don't technically need it, I was going to use it as the collection key in the UI?

**Comment by davidackroyd99** [comment]
Think I've resolved all your comments, thanks for being so helpful 👍

File: src/BaGet.Azure/Search/AzureSearchService.cs [hunk 7]
```diff
@@ -163,7 +163,11 @@ public class AzureSearchService : ISearchService
 
             var response = await _searchClient.Documents.SearchAsync<PackageDocument>(query, parameters, cancellationToken: cancellationToken);
             var results = response.Results
-                .Select(r => r.Document.Id)
+                .Select(r => new DependentResult() {
```
Comment (loic-sharma) [inline src/BaGet.Azure/Search/AzureSearchService.cs:166]: Minor style nitpick:

```suggestion
                .Select(r => new DependentResult
                {
```

File: src/BaGet.Core/Search/DatabaseSearchService.cs [hunk 8]
```diff
@@ -125,7 +125,12 @@ public async Task<DependentsResponse> FindDependentsAsync(string packageId, Canc
                 .OrderByDescending(p => p.Downloads)
                 .Where(p => p.Dependencies.Any(d => d.Id == packageId))
                 .Take(20)
-                .Select(p => p.Id)
+                .Select(r => new DependentResult()
```
Comment (loic-sharma) [inline src/BaGet.Core/Search/DatabaseSearchService.cs:128]: Minor style nitpick:

```suggestion
                .Select(r => new DependentResult
```

**Comment by loic-sharma** [comment]
Thanks for the contribution, this is awesome! 🥳

**Pull request merged**